File: frontend/src/types/area-selection.ts
/**
 * Area Selection Types for Tiling Feature
 *
 * Coordinates are relative to the PDF page at native resolution (e.g., 600 DPI).
 */

/**
 * Represents a rectangular area selection on a PDF page.
 */
export interface AreaSelection {
  pageNum: number; // 1-indexed page number
  x: number; // Left edge in pixels (native DPI)
  y: number; // Top edge in pixels (native DPI)
  width: number; // Width in pixels (native DPI)
  height: number; // Height in pixels (native DPI)
}

/**
 * State for managing area selections across pages.
 */
export interface AreaSelectionState {
  selections: Map<number, AreaSelection>; // pageNum -> selection
  activePageNum: number | null; // Currently drawing on this page
}

/**
 * Pagination state for page preview.
 */
export interface PagePreviewPagination {
  currentStartPage: number; // 1-indexed, first page being displayed
  pagesPerView: 2; // Always 2 (constant)
  totalPages: number;
}

/**
 * Estimated tile count for an area selection.
 */
export interface TileEstimate {
  pageNum: number;
  rows: number;
  cols: number;
  total: number;
}

/**
 * Page dimensions for coordinate transformation.
 */
export interface PageDimensions {
  nativeWidth: number; // PDF width at target DPI (e.g., 600)
  nativeHeight: number;
  displayWidth: number; // Rendered canvas width
  displayHeight: number;
}

/**
 * Tile preview from estimate endpoint.
 */
export interface TilePreview {
  row: number; // 0-indexed
  col: number; // 0-indexed
  x: number; // Pixel x of tile top-left
  y: number; // Pixel y of tile top-left
}

/**
 * Response from tile estimate endpoint.
 */
export interface TileEstimateResponse {
  rows: number;
  cols: number;
  total: number;
  tiles: TilePreview[];
}

/**
 * Request for tile estimate endpoint.
 */
export interface TileEstimateRequest {
  pageWidth: number;
  pageHeight: number;
  tileSize?: number;
  overlap?: number;
  areaSelection?: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
}

File: src/models/area_selection.py
from dataclasses import dataclass
from typing import List


@dataclass
class AreaSelection:
    """Rectangular area selection on a PDF page.

    Coordinates are relative to the PDF page at native resolution (e.g., 600 DPI).
    """
    page_num: int      # 1-indexed page number
    x: int             # Left edge in pixels (native DPI)
    y: int             # Top edge in pixels (native DPI)
    width: int         # Width in pixels
    height: int        # Height in pixels

    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization."""
        return {
            'pageNum': self.page_num,
            'x': self.x,
            'y': self.y,
            'width': self.width,
            'height': self.height
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'AreaSelection':
        """Create from dictionary (e.g., from JSON request)."""
        return cls(
            page_num=data['pageNum'],
            x=data['x'],
            y=data['y'],
            width=data['width'],
            height=data['height']
        )

    def clamp_to_bounds(self, page_width: int, page_height: int) -> 'AreaSelection':
        """Return a new AreaSelection clamped to page boundaries."""
        x = max(0, min(self.x, page_width - 1))
        y = max(0, min(self.y, page_height - 1))
        width = min(self.width, page_width - x)
        height = min(self.height, page_height - y)
        return AreaSelection(
            page_num=self.page_num,
            x=x,
            y=y,
            width=max(1, width),
            height=max(1, height)
        )

    @property
    def right(self) -> int:
        """Right edge x coordinate."""
        return self.x + self.width

    @property
    def bottom(self) -> int:
        """Bottom edge y coordinate."""
        return self.y + self.height
